File: gems/persona/lib/persona/oidc.rb
require 'cgi'

module Persona
  # OIDC account linking — the verification half of binding a persona to an
  # external identity provider. persona-protocol owns the flow (doc/protocol.md
  # §7); the application registers one provider object per IdP via
  # Persona.config.register_oidc_provider, and gains new providers by
  # updating this gem and registering them — not by reimplementing the flow.
  #
  # A provider is any object exposing:
  #
  #   name              -> String   stable identifier stored on account
  #                                 bindings ('google', 'uniba-auth', ...)
  #   authorize(state)  -> AuthorizeStart   the authorize URL plus any per-flow
  #                                         secrets (PKCE verifier, nonce) to
  #                                         stash with the state
  #   verify(params, stash = nil) -> Identity | nil   callback verification,
  #                                                   using the stashed secrets
  #
  # Persona::Oidc::Verifier is the production implementation (code exchange +
  # JWKS validation); StubProvider is the local stand-in.
  module Oidc
    # The verified outcome of an OIDC round-trip: which provider vouched for
    # the browser, and the stable subject the token was issued for. This
    # pair becomes an account binding.
    Identity = Struct.new(:provider, :subject, keyword_init: true)

    # What #authorize returns: where to send the browser, and the per-flow
    # secrets to stash with the state and hand back to #verify on the
    # callback. The stub needs no stash; a real code+PKCE flow does.
    AuthorizeStart = Struct.new(:url, :stash, keyword_init: true)

    # Local / development stand-in for a real provider. Its authorize URL
    # points at an app-local page, and #verify trusts a subject handed
    # straight through the callback params — NO cryptographic verification.
    # It exists so the whole bind / adopt / merge flow can be exercised
    # before any real IdP is wired up. MUST NOT be deployed to production
    # (protocol.md P-16).
    class StubProvider
      attr_reader :name

      def initialize(name: 'stub', authorize_path: '/auth/oidc/start')
        @name = name
        @authorize_path = authorize_path
      end

      def authorize(state)
        AuthorizeStart.new(url: "#{@authorize_path}?state=#{CGI.escape(state)}", stash: nil)
      end

      def verify(params, _stash = nil)
        subject = params[:sub] || params['sub']
        return nil if subject.nil? || subject.to_s.empty?

        Identity.new(provider: name, subject: subject)
      end
    end
  end
end

File: packages/server-core/src/oidc.ts
// OIDC account linking — the provider contract and registry (doc/protocol.md
// §7). A provider is one object per IdP behind three methods; applications gain
// a new provider by registering it, not by reimplementing the flow. Mirrors the
// Ruby gem's Persona::Oidc (gems/persona/lib/persona/oidc.rb).

// The verified outcome of an OIDC round-trip: which provider vouched for the
// browser, and the stable subject the token was issued for. This pair becomes
// an account binding.
export type Identity = { provider: string; subject: string };

export interface OidcProvider {
  // Stable identifier stored on account bindings ('google', 'my-org-idp', …).
  readonly name: string;
  // Where the begin step sends the browser to authenticate. Only the opaque
  // state may ride the URL (P-14).
  authorizeUrl(state: string): string;
  // Callback verification. Real providers exchange the code and validate the
  // token against the IdP's published keys; may be async.
  verify(params: Record<string, unknown>): Identity | null | Promise<Identity | null>;
}

// Local / development stand-in for a real provider. Its authorize URL points at
// an app-local page, and verify trusts a subject handed straight through the
// callback params — NO cryptographic verification. MUST NOT be deployed to
// production (P-16).
export class StubProvider implements OidcProvider {
  readonly name: string;
  private readonly authorizePath: string;

  constructor(opts: { name?: string; authorizePath?: string } = {}) {
    this.name = opts.name ?? 'stub';
    this.authorizePath = opts.authorizePath ?? '/auth/oidc/start';
  }

  authorizeUrl(state: string): string {
    return `${this.authorizePath}?state=${encodeURIComponent(state)}`;
  }

  verify(params: Record<string, unknown>): Identity | null {
    const sub = params.sub;
    if (typeof sub !== 'string' || sub === '') return null;
    return { provider: this.name, subject: sub };
  }
}

export interface ProviderRegistry {
  register(provider: OidcProvider): OidcProvider;
  // Resolves a provider by name; throws for a name that was never registered so
  // a begin/callback for an unknown provider fails loudly (mirrors the Ruby
  // Persona.oidc_provider contract).
  get(name: string): OidcProvider;
  has(name: string): boolean;
}

// Empty by default: register providers explicitly, including the development
// stub — an unregistered provider can never verify.
export const createProviderRegistry = (providers: OidcProvider[] = []): ProviderRegistry => {
  const map = new Map<string, OidcProvider>();
  for (const p of providers) map.set(p.name, p);
  return {
    register(provider) {
      map.set(provider.name, provider);
      return provider;
    },
    get(name) {
      const provider = map.get(name);
      if (!provider) {
        throw new Error(`unknown OIDC provider ${JSON.stringify(name)} — register it via the provider registry`);
      }
      return provider;
    },
    has(name) {
      return map.has(name);
    },
  };
};
